File: skybit/hud.py
import pyxel
from .config import TRIPLE_DURATION
from .palette import INK, CREAM, GOLD, ORNG, NAVY, WHITE, DKRED, GREEN, RED
from .sprites import draw_text, draw_text_c, text_w, draw_mushroom, draw_coin


def draw_hud(world, high_score):
    W = pyxel.width

    # Score (large, top-centre)
    s  = str(world.score)
    sw = text_w(s, 2)
    draw_text(s, (W - sw) // 2, 8, WHITE, 2, shadow_col=INK)

    # High score (top-left)
    draw_text(f'HI {high_score}', 4, 4, CREAM, 1, shadow_col=INK)

    # Coin count (top-right, with tiny coin icon)
    cs  = f'x{world.coin_count}'
    cw  = text_w(cs)
    cx  = W - cw - 14
    pyxel.circ(W - cw - 10, 6, 3, GOLD)
    pyxel.pset(W - cw - 11, 5, CREAM)
    draw_text(cs, cx + 3, 4, GOLD, 1, shadow_col=INK)

    # Pause button (top-right)
    px, py = W - 14, 14
    pyxel.rect(px, py, 11, 11, INK)
    pyxel.rect(px + 1, py + 1, 9, 9, NAVY)
    pyxel.rect(px + 3, py + 3, 2, 5, CREAM)
    pyxel.rect(px + 6, py + 3, 2, 5, CREAM)

    # Triple-score timer bar
    if world.triple_t > 0:
        bw = 70; bh = 4
        bx = (W - bw) // 2; by = 24
        pyxel.rect(bx - 1, by - 1, bw + 2, bh + 2, INK)
        pyxel.rect(bx, by, bw, bh, NAVY)
        fill = max(0, min(bw, int(bw * world.triple_t / TRIPLE_DURATION)))
        pyxel.rect(bx, by, fill, bh, ORNG)
        draw_text_c('3X POWER', by - 8, CREAM, 1, shadow_col=INK)

    # Combo badge
    if world.combo >= 2:
        lbl = f'X{world.combo} COMBO!'
        draw_text_c(lbl, pyxel.height - 28, ORNG, 1, shadow_col=INK)

    # Floating "+1"/"+3" texts
    for f in world.floats:
        lw = text_w(f.text)
        draw_text(f.text, int(f.x) - lw//2, int(f.y), f.col, 1, shadow_col=INK)

    # Pause-button hit rect returned for input
    return (px, py, 11, 11)


def draw_title(t):
    W = pyxel.width
    bob = int(2 * pyxel.sin(t * 60 * 0.4))   # pyxel.sin is degree-based
    y   = 52 + bob

    # Shadow
    draw_text('SKYBIT', (W - text_w('SKYBIT', 3))//2 + 2, y + 2, INK, 3)
    # Gold layer
    draw_text('SKYBIT', (W - text_w('SKYBIT', 3))//2 + 1, y - 1, GOLD, 3)
    # Orange layer
    draw_text('SKYBIT', (W - text_w('SKYBIT', 3))//2,     y,     ORNG, 3)

    draw_text_c('A RETRO PIXEL FLYER', y + 24, CREAM, 1, shadow_col=INK)

    # Preview parrot + coins
    from .sprites import draw_parrot_tilt, draw_coin
    draw_parrot_tilt(40, 155, int(t * 4) % 4, -50)
    draw_coin(85,  158, t)
    draw_coin(100, 152, t + 1.0)
    draw_mushroom_small(130, 162, t)


def draw_mushroom_small(x, y, t):
    from .sprites import draw_mushroom
    draw_mushroom(x, y, t)


def draw_start_hint():
    W = pyxel.width
    y1, y2 = 178, 194
    draw_text_c('TAP TO FLAP',             y1, WHITE, 1, shadow_col=INK)
    draw_text_c('COLLECT COINS',           y2, GOLD,  1, shadow_col=INK)
    draw_text_c('GRAB MUSHROOM FOR 3X!', y2+12, ORNG,  1, shadow_col=INK)


def draw_game_over(score, high, new_best):
    W = pyxel.width
    y = 82
    draw_text_c('GAME OVER', y,     DKRED, 2, shadow_col=INK)
    draw_text_c(f'SCORE {score}',   y+24, WHITE, 1, shadow_col=INK)
    draw_text_c(f'BEST  {high}',    y+36, GOLD,  1, shadow_col=INK)
    if new_best:
        draw_text_c('NEW BEST!',    y+50, ORNG,  1, shadow_col=INK)
    draw_text_c('TAP TO RETRY',     y+70, CREAM, 1, shadow_col=INK)


def draw_paused():
    # Dim the screen
    pyxel.dither(0.5)
    pyxel.rect(0, 0, pyxel.width, pyxel.height, INK)
    pyxel.dither(1.0)
    draw_text_c('PAUSED',     pyxel.height//2 - 12, WHITE, 2, shadow_col=INK)
    draw_text_c('TAP RESUME', pyxel.height//2 + 10, CREAM, 1, shadow_col=INK)

File: src/hud.js
import { VIEW_W, VIEW_H, TRIPLE_DURATION } from './config.js';
import { C } from './palette.js';
import { drawText, drawTextCentered, measureText } from './sprites.js';

export function drawHUD(buf, world, highScore) {
    // Score (big, top-center)
    const s = String(world.score);
    const w = measureText(s, 2);
    const sx = ((VIEW_W - w) / 2) | 0;
    drawText(buf, s, sx + 1, 11, C.NIGHT, 2);
    drawText(buf, s, sx, 10, C.WHITE, 2);

    // High score (top-left)
    const hs = 'HI ' + String(highScore);
    drawText(buf, hs, 6, 6, C.NIGHT);
    drawText(buf, hs, 5, 5, C.CREAM);

    // Coins (top-right)
    const cs = 'x' + String(world.coinCount);
    const cw = measureText(cs);
    drawText(buf, cs, VIEW_W - cw - 7, 6, C.NIGHT);
    drawText(buf, cs, VIEW_W - cw - 8, 5, C.GOLD);
    // tiny coin icon next to count
    buf.disc(VIEW_W - cw - 12, 7, 2, C.GOLD);
    buf.put(VIEW_W - cw - 12, 6, C.CREAM);

    // Pause button (top-right corner, below coin)
    const pr = { x: VIEW_W - 16, y: 16, w: 12, h: 12 };
    buf.rect(pr.x, pr.y, pr.w, pr.h, C.NIGHT);
    buf.rect(pr.x + 1, pr.y + 1, pr.w - 2, pr.h - 2, C.DUSK);
    buf.rect(pr.x + 3, pr.y + 3, 2, 6, C.CREAM);
    buf.rect(pr.x + 7, pr.y + 3, 2, 6, C.CREAM);

    // Triple power timer bar.
    if (world.tripleTimer > 0) {
        const bw = 80, bh = 5;
        const bx = ((VIEW_W - bw) / 2) | 0;
        const by = 26;
        buf.rect(bx - 1, by - 1, bw + 2, bh + 2, C.NIGHT);
        buf.rect(bx, by, bw, bh, C.NAVY);
        const fill = Math.max(0, Math.min(1, world.tripleTimer / TRIPLE_DURATION));
        buf.rect(bx, by, (bw * fill) | 0, bh, C.ORANGE);
        const label = '3X POWER';
        drawText(buf, label, ((VIEW_W - measureText(label)) / 2) | 0, by - 7, C.CREAM);
    }

    // Combo badge (bottom-center when combo >= 2)
    if (world.combo >= 2) {
        const label = 'X' + world.combo + ' COMBO';
        const lw = measureText(label);
        const ly = 270;
        const lx = ((VIEW_W - lw) / 2) | 0;
        drawText(buf, label, lx + 1, ly + 1, C.NIGHT);
        drawText(buf, label, lx, ly, C.ORANGE);
    }

    // Float texts (coin pickup "+1")
    for (const t of world.floatTexts) {
        const lw = measureText(t.text);
        const fx = (t.x - lw / 2) | 0;
        const fy = t.y | 0;
        drawText(buf, t.text, fx + 1, fy + 1, C.NIGHT);
        drawText(buf, t.text, fx, fy, t.color);
    }

    return pr; // report pause-button hit rect for input.
}

export function drawStartHint(buf) {
    const y1 = 180, y2 = 196;
    drawTextCentered(buf, 'TAP TO FLAP', y1, C.NIGHT);
    drawTextCentered(buf, 'TAP TO FLAP', y1 - 1, C.WHITE);
    drawTextCentered(buf, 'COLLECT COINS', y2, C.NIGHT);
    drawTextCentered(buf, 'COLLECT COINS', y2 - 1, C.GOLD);
    drawTextCentered(buf, 'GRAB THE MUSHROOM FOR 3X!', y2 + 12, C.NIGHT);
    drawTextCentered(buf, 'GRAB THE MUSHROOM FOR 3X!', y2 + 11, C.ORANGE);
}

export function drawTitle(buf, t) {
    // Animated title: bounces a bit.
    const bob = Math.round(Math.sin(t * 2.4) * 2);
    const y = 64 + bob;
    const title = 'SKYBIT';
    const w = measureText(title, 3);
    const x = ((VIEW_W - w) / 2) | 0;
    drawText(buf, title, x + 2, y + 2, C.NIGHT, 3);
    drawText(buf, title, x, y, C.ORANGE, 3);
    drawText(buf, title, x + 1, y - 1, C.GOLD, 3);

    const sub = 'A RETRO PIXEL FLYER';
    drawTextCentered(buf, sub, y + 24, C.NIGHT);
    drawTextCentered(buf, sub, y + 23, C.CREAM);
}

export function drawGameOver(buf, score, high, best) {
    const y = 100;
    const t = 'GAME OVER';
    const w = measureText(t, 2);
    const x = ((VIEW_W - w) / 2) | 0;
    drawText(buf, t, x + 2, y + 2, C.NIGHT, 2);
    drawText(buf, t, x, y, C.CRIMSON, 2);

    drawTextCentered(buf, 'SCORE ' + score, y + 24, C.NIGHT);
    drawTextCentered(buf, 'SCORE ' + score, y + 23, C.WHITE);
    drawTextCentered(buf, 'BEST  ' + high, y + 38, C.NIGHT);
    drawTextCentered(buf, 'BEST  ' + high, y + 37, C.GOLD);

    if (best) {
        drawTextCentered(buf, 'NEW BEST!', y + 54, C.NIGHT);
        drawTextCentered(buf, 'NEW BEST!', y + 53, C.ORANGE);
    }

    drawTextCentered(buf, 'TAP TO RETRY', y + 82, C.NIGHT);
    drawTextCentered(buf, 'TAP TO RETRY', y + 81, C.CREAM);
}
